### PR TITLE
fix: Kl/h01-gw-audit-2

### PR DESCRIPTION
**Pull request by kelemeno**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol [hunk 1]
```diff
@@ -359,42 +359,54 @@ abstract contract NativeTokenVault is INativeTokenVault, IAssetHandler, Ownable2
...
-        bytes32 _assetId,
-        address _originToken
-    ) internal view returns (address expectedToken) {
-        expectedToken = calculateCreate2TokenAddress(_originChainId, _originToken);
-        bytes32 expectedAssetId = DataEncoding.encodeNTVAssetId(_originChainId, _originToken);
+    function _calculateExpectedTokenAddress(
+        address _originToken,
+        bytes memory _erc20Data
+    ) internal view returns (address expectedToken, uint256 tokenOriginChainId) {
+        tokenOriginChainId = this.tokenDataOriginChainId(_erc20Data);
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol:381]: why is this an external call?

File: l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol [hunk 2]
```diff
@@ -127,13 +121,34 @@ contract L2NativeTokenVault is IL2NativeTokenVault, NativeTokenVault {
...
     /// @notice Deploys the beacon proxy for the L2 token, while using ContractDeployer system contract.
     /// @dev This function uses raw call to ContractDeployer to make sure that exactly `l2TokenProxyBytecodeHash` is used
     /// for the code of the proxy.
     /// @param _salt The salt used for beacon proxy deployment of L2 bridged token.
     /// @return proxy The beacon proxy, i.e. L2 bridged token.
-    function _deployBeaconProxy(bytes32 _salt) internal virtual override returns (BeaconProxy proxy) {
-        if (address(L2_LEGACY_SHARED_BRIDGE) == address(0)) {
+    function _deployBeaconProxy(
+        bytes32 _salt,
+        uint256 _tokenOriginChainId
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol:149]: doc comments are no longer valid

File: l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol [hunk 3]
```diff
@@ -127,13 +121,34 @@ contract L2NativeTokenVault is IL2NativeTokenVault, NativeTokenVault {
...
+    ) internal virtual override returns (BeaconProxy proxy) {
+        if (address(L2_LEGACY_SHARED_BRIDGE) == address(0) && _tokenOriginChainId != L1_CHAIN_ID) {
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol:151]: this looks weird, if L2_LEGACY_SHARED_BRIDGE) == address(0) and _tokenOriginChainId == L1_CHAIN_ID you will go into the `else` statement and use `L2_LEGACY_SHARED_BRIDGE`

File: l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol [hunk 4]
```diff
@@ -416,28 +428,20 @@ abstract contract NativeTokenVault is INativeTokenVault, IAssetHandler, Ownable2
...
-        if (block.chainid == L1_CHAIN_ID && tokenOriginChainId == 0) {
-            revert L1TokenDeploymentWithZeroChainId(_assetId);
-        }
-        tokenOriginChainId = tokenOriginChainId == 0 ? L1_CHAIN_ID : tokenOriginChainId;
-        if (tokenOriginChainId == block.chainid) {
+        bytes32 salt = _getCreate2Salt(_tokenOriginChainId, _originToken);
+
+        BeaconProxy l2Token = _deployBeaconProxy(salt, _tokenOriginChainId);
+        BridgedStandardERC20(address(l2Token)).bridgeInitialize(_assetId, _originToken, _erc20Data);
+        if (_tokenOriginChainId == block.chainid) {
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol:440]: we can do this check at the top of the function for failing faster/wasting less gas

**Comment by StanislavBreadless** [comment]
Also, all the new functions need to be properly doc commented

File: l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol (hunk 1 above)
Comment (kelemeno) [reply l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol:381]: erc20 data is calldata in decode

File: l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol (hunk 3 above)
Comment (kelemeno) [reply l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol:151]: true, its ||

**Pull request closed without merging**